### ISSUE TITLE
Enable Terminal to persist / restore instance settings

### ISSUE THREAD

**Issue by zadjii-msft** [microsoft/terminal#766]
Let the user set a global setting to have the terminal re-open with whatever profiles were last open.

Thoughts:
 * what happens when a user has two windows open, closes one with 3 tabs, then opens a new window? Should it open the 3 tabs? Or should it start a fresh, single tabbed window?
 * how do we persist the last session?
 * Do we only persist the last tabs open when the last terminal window is closed? 
 * Do we add a "quit terminal" option that closes all open windows, and persists all their open profiles?

__THIS IS NOT A DISCUSSION OF PRESERVING WINDOW *CONTENT*, ONLY OPEN WINDOW PROFILES__

This also includes restoring things like previous window positions.

**Comment by glen-84** [comment]
1. Will this also restore split pane configurations?
2. Will this restore the current directory (CWD/PWD) of each tab (as opposed to just the configured starting directory)?

---

IMO:

> what happens when a user has two windows open, closes one with 3 tabs, then opens a new window? Should it open the 3 tabs? Or should it start a fresh, single tabbed window?

Open a fresh window. If you later add support for window history, they could use a "Recently closed windows" menu to restore the window with the same set of tabs.

> how do we persist the last session?
> Do we only persist the last tabs open when the last terminal window is closed?

That would be the simplest solution probably, but it might not be ideal if the application is improperly shut down. An alternative would be to continuously track open windows, tabs, panes, and working directories as changes are made, and persist those changes immediately (or on a timer).

> Do we add a "quit terminal" option that closes all open windows, and persists all their open profiles?

Firefox and Chrome have an "Exit" option, which closes all windows. When you start the application again, the windows are restored.

In general, I think that it would be great if Windows Terminal mimicked the behaviour of browsers.

**Comment by mdtauk** [comment]
I know this is just about tabs and pane arrangements, but why not restore the buffer with all the previous output strings in place?

**Comment by glen-84** [comment]
I was thinking, if the state is stored in a human-readable format, it could also be hand-edited by users, and generated/modified by an external script or programme.

A made-up example:

```jsonc
{
    "version": "0.2.0",
    "state": {
        "windows": [
            {
                "maximized": false,
                "positionX": 500,
                "positionY": 200,
                "width": 1000,
                "height": 2000,
                "tabs": [
                    {
                        "title": "Project X",
                        "panes": [
                            {
                                "profile": "EF3E32A7-5FF6-42B4-B6E2-96CD7D033F00",
                                "directory": "C:/...",
                                // position/dimensions
                            }
                        ]
                    }
                ]
            }
        ]
    }
}
```

(As usual, I wish that people would move away from JSON and instead use something more suited to configuration files, like YAML.)

---

@mdtauk,

> I know this is just about tabs and pane arrangements, but why not restore the buffer with all the previous output strings in place?

To limit the scope of the feature and get it released sooner. The format that I show above could easily be extended at a later stage.

**Comment by ExE-Boss** [comment]
The thing that makes me hate YAML is that the spec doesn’t support tabs.

**Comment by JCKodel** [comment]
* what happens when a user has two windows open, closes one with 3 tabs, then opens a new window? Should it open the 3 tabs? Or should it start a fresh, single tabbed window?
Use the browser behavior. It works and it is familiar. Also, would be nice to detach/reatach tabs (as Chrome).

* how do we persist the last session?
It would be nice to persist everything (including history) of each tab. It would be amazing to have up arrow history persisted as well. You can use all my %USERPROFILE% space for that =)

* Do we only persist the last tabs open when the last terminal window is closed?
As in a browser, when a tab is closed, its state remains (so I can reopen closed tabs with Ctrl+Shift+T). This would be awesome. Perhaps some settings to limit the number of persisted sessions?

* Do we add a "quit terminal" option that closes all open windows, and persists all their open profiles?
As in a browser, [X] closes all tabs (I can reopen them with Ctrl+Shift+T - it will restore entire windows if needed).

**Comment by bitcrazed** [comment]
Updated title to better describe the ask, which also came-up in this Twitter thread: https://twitter.com/stanzillaz/status/1176675911275556866

**Comment by DHowett-MSFT** [comment]
Renaming it _again_ because it really seems to pertain to saving info about specific instances of specific profiles. We can quibble about session/profile/instance later. :smile:

**Comment by getify** [comment]
Still feeling this pain every time I re-open. I don't even mind it not remembering size, since I set that in my config. I just really hate it not remembering position, because on my system, every time it opens, the top-left corner is about 30px to the right/down from the last time it opened, so it shifts little by little until something resets it back to the top-left and the cycle repeats. That is endlessly annoying.

Would love even a minimal interim on window size being remembered quicker, even if this larger feature is still being worked out. Please!? :)

**Comment by zadjii-msft** [comment]
@getify You know you can use the [`initialPosition`](https://docs.microsoft.com/en-us/windows/terminal/customize-settings/global-settings#launch-position) to set a launch position for the Terminal window, right?

**Comment by getify** [comment]
@zadjii-msft No, I was not aware of that. Maybe it was added after the last time I had looked at the docs.

Just tried it, and with some fiddling it seems fine for now as an interim. I use a left-side taskbar, so it was a little annoying that the "X" position isn't "available desktop coordinates" but "absolute screen position". I figured out with some trial and error that `93` was the right magic number for my current setup.

**Comment by smykhailov** [comment]
I found some workaround for this, which works for me and maybe will be helpful for others.
I created cmd file with the following command:
`wt.exe -p PowerShell -d "D:\Projects" ; split-pane -p PowerShell -d "D:\Projects\foo" ; split-pane -H -p PowerShell -d "D:\Projects\bar" ; -p cmd -d "D:\Projects" ; -p Ubuntu ; focus-tab -t 0`
which opens the tabs and panes I need to have opened.

also in `settings.json` I added the position and size:
```JSON
  "initialPosition": "-1760, 20",
  "initialRows": 46,
  "initialCols": 170,
```
with `x: -1760` I open terminal on my left monitor :)

**Comment by mws-rmain** [comment]
When restoring 'last position', please remember to handle which monitor (assuming more than one) the position refers to.

**Comment by igorakkerman** [comment]
Do not only remember the last monitor that was used, but also remember the last position used on other monitors

When I have the Terminal open on my big monitor and disconnect my notebook, it will appear on my notebook screen, of course. Now, it would be great if Terminal remembered the last position used when it was open on my notebook screen. Then, when I would re-connect my big monitor, the Terminal should appear on my big monitor, remembering the last position on that screen.

**Comment by radutomy** [comment]
Also remember position if opened on another virtual desktop.

**Comment by clounie** [comment]
> Do not only remember the last monitor that was used, but also remember the last position used on other monitors.
> 
> When I have the Terminal open on my big monitor and disconnect my notebook, it will appear on my notebook screen, of course. Now, it would be great if Terminal remembered the last position used when it was open on my big monitor's _and_ my notebook's screens. Then, when I would re-connect my big monitor, the Terminal should appear on my big monitor, at the last position on that screen. If I disconnect my big monitor again, the Terminal should come back to the last position on my notebook's screen.
> 
> That is, the application should store the last monitor used and a table mapping each monitor to the position and dimensions the Terminal occupied on that monitor.

^ This would be great. But would rather have basic functionality than wait extra time for macOS tricks.

@glen-84 Pretty much nailed behavior with his comment about "[do as browsers do](https://github.com/microsoft/terminal/issues/766#issuecomment-494567076)". It keeps things fairly simple, and those are patterns that, while fairly powerful, anyone who uses a computer is familiar with.

**Comment by NeilMacMullen** [comment]
As described in #8989, if the state can be persisted to file as suggested in https://github.com/microsoft/terminal/issues/766#issuecomment-494689538 then a desirable feature would be to be able to invoke a new terminal instance and supply that *particular* configuration as the startup state (via argument).  That would help alleviate some of the ambiguity around what to do in the case of multiple open instances.

**Comment by brantPTS** [comment]
> Let the user set a global setting to have the terminal re-open with whatever profiles were last open.
> 
> Thoughts:
> 
> * what happens when a user has two windows open, closes one with 3 tabs, then opens a new window? Should it open the 3 tabs? Or should it start a fresh, single tabbed window?
> * how do we persist the last session?
> * Do we only persist the last tabs open when the last terminal window is closed?
> * Do we add a "quit terminal" option that closes all open windows, and persists all their open profiles?
> 
> **THIS IS NOT A DISCUSSION OF PRESERVING WINDOW _CONTENT_, ONLY OPEN WINDOW PROFILES**

@zadjii-msft 

You should have this restore just like a browser, i.e., either restore last window closed, or if they select "Exit" command, restore all windows. What would be really awesome would be to save command history in each tab so you can just up arrow to bring back recent commands. That will make this indispensable. 

Thank you.

**Comment by zadjii-msft** [comment]
> What would be really awesome would be to save command history in each tab so you can just up arrow to bring back recent commands.

FYI this has been discussed to death elsewhere in this repo - it's not the terminal's responsibility to remember what the last commands entered were. Most modern shells (e.g. `pwsh`, `bash`, `zsh`) will already do this. `cmd` can't do this, because `cmd` has never done that, and `cmd` is in maintenance-only mode. 

The rest though - I'm on board with all that ☺️

**Comment by Poopooracoocoo** [comment]
hey @zadjii-msft (sorry if you're not back from your vacay) this issue doesn't sound like remembering the size and position of a Terminal window to many users, including myself. The discussion below it and the closed issues do make it look like it, but not everyone will have the time and patience to do that. Mind updating the issue to say something about it _independent_ of there being an option to restore the last session? :)

**Comment by zadjii-msft** [comment]
Okay, sure. I'll update the OP. To clarify:
* #766 (This issue) is tracking restoring the relative window state across terminal launches. This includes (but is not limited to) window positions, the open tab/pane structure, and what profiles were in those tab/panes. It is _not_ tracking restoring buffer state. 
* #961 is specifically related to trying to restore the _buffer_ content of these tabs/panes. This is substantially harder than just restoring the window state, because the buffer has quite a bit more state in it. It is _not_ tracking connection state (commandline history, environment variables, open processes)
* #960 is about just "restoring previously opened tabs/panes". This is like when you close a tab in a browser, then hit <kbd>Ctrl+shift+t</kbd> to just pop it back into existence. This does not include across reboots - it mainly deals with tab/panes closed in the current instance.
* There's not an issue tracking restoring connection/shell/process state, because that's _generally impossible_. (case in point: https://github.com/microsoft/terminal/issues/961#issuecomment-495506932, et. al in that thread.)

**Comment by igorakkerman** [comment]
@zadjii-msft thanks for clarifying. In your first item, you refer to restoring the window state _across_ terminal launches. Should we have another issue for restoring the window state _while it is running_, but after an external monitor is disconnected and re-connected? (https://github.com/microsoft/terminal/issues/766#issuecomment-653671768)

**Comment by zadjii-msft** [comment]
That's totally something that should have it's own issue. FYI I don't think _any other app_ does this currently - I'm constantly repositioning windows when I dock/undock my laptop.

**Comment by marcelo-mason** [comment]
This issue is opened for two years.  Unusable terminal for any poweruser for two years.  How about just get something done, then figure out how to make it perfect later.

**Comment by yodaldevoid** [comment]
@marcelo-mason If you feel so strongly about it, as it is an open source project you are always free to submit a proposed spec for the feature. While I don't know for certain, I have the feeling that the obvious movement on this feature is not due to people not thinking it is important but because other features have taken precedence because people have decided that they are more important. This feature might see more movement if it had a champion to push it forward; maybe you fit that roll?

**Comment by zadjii** [comment]
That's absolutely true. This is a big project, with a lot of high priority features in development all the time. A lot of the groundwork for this particular feature is in progress over at #8771. There's just only so much bandwidth for designing so many features at a time, and the team is currently busy with the Settings UI (#1564), quake mode (#653), tab tear out (#1256), setting a default terminal application (#7489), so there's not really any other bandwidth at the moment. It's still high up on the priority list though, and if anyone would like to make progress faster, I'm sure the team would be happy to review a spec proposal or a PR ☺️

**Comment by zadjii-msft** [comment]
> Terminal STILL doesn't persist its window size and position!

Did you enable the setting? 
![image](https://user-images.githubusercontent.com/18356694/138146770-7b4c31e3-586f-4660-bed3-15d4b6d6f5a3.png)
It won't work for elevated windows at the moment.

> And strangely it doesn't seem to use WinUI 2.7 with the new styles.

That's by design, see #11231

> Something that's present in the release build is a strange window that is opened briefly before Terminal opens.

File a new issue for that, let's not take this thread off topic.

**Comment by Rosefield** [comment]
I apologize if the naming is confusing. The issue, I think, is that a "session" is not well defined yet anywhere, and the wording was based on other applications (such as Edge) to be forward looking for when things like #961 are also implemented.

The exact behavior of this feature when the setting is enabled is, from the docs (https://docs.microsoft.com/en-us/windows/terminal/customize-settings/startup#behavior-when-starting-a-new-terminal-session-preview)
```
Note: [when enabled] Currently, Windows Terminal will save the following information:
- Each window's position, size, and name
- The tab layout of each window, including the layout and profile of each pane, but not any contents of those panes
```

**Comment by Poopooracoocoo** [comment]
>That being said, people FUCKING HATE defaults being changed on them, so no promises.

[Insert XKCD]

but i honestly expected that behaviour to not be behind _both_ a preview release and an opt-in. 
the very question of whether it becomes default or not makes me sad. it just makes sense to match standard window conventions :/ 
like i've been waiting for like two years here :'( but hopefully it'll be there in 1.12!!! ^.^

**Comment by Rosefield** [comment]
Since you're already using 1.12 preview you can use this today. Open your settings (press `ctr+,`) and then make sure the settings is enabled. Once it is enabled windows will be saved when you use the `quit` action, or by pressing the `X` icon on the window.

![image](https://user-images.githubusercontent.com/6185249/138161469-77d9b8da-3d87-4b24-9331-54091520ad1c.png)

**Comment by igorakkerman** [comment]
This works for me when launching Terminal regularly. However, when I launch it as an administrator, neither does it open the last session I had as a regular user, nor does it open the last administrator session or even restore the window's position and size.

**Comment by rehanhashmi** [comment]
> ```js
> "state": {
>         "windows": [
>             {
>                 "maximized": false,
>                 "positionX": 500,
>                 "positionY": 200,
>                 "width": 1000,
>                 "height": 2000,
>                 "tabs": [
>                     {
>                         "title": "Project X",
>                         "panes": [
>                             {
>                                 "profile": "EF3E32A7-5FF6-42B4-B6E2-96CD7D033F00",
>                                 "directory": "C:/...",
>                                 // position/dimensions
>                             }
>                         ]
>                     }
>                 ]
>             }
>         ]
>     }
> ```

Even this one not working for me, I tried and reopen terminal but is not working.

**Comment by zadjii-msft** [comment]
@rehanhashmi did you write that json by hand? That doesn't look like the actual schema we ended up going with for the state file.

Make sure to enable `"firstWindowPreference": "persistedWindowLayout"` in the settings, and the Terminal will auto-generate state that looks more like:
![image](https://user-images.githubusercontent.com/18356694/146177991-04dc3390-f153-4c95-8576-de100c10ff20.png)

in `state.json`.

> Tabs names ... are lost

That's a known bug, #11878

>  directories are lost

You'll need to make sure to [configure your shell](https://docs.microsoft.com/en-us/windows/terminal/tutorials/new-tab-same-directory) to emit the CWD to the Terminal for us to be able to restore those

**Comment by rehanhashmi** [comment]
@zadjii-msft Here is my settings.json file from windows shell, My requirement is like each time mostly I need to move to some directory and run few commands. Each time I open terminal I make few tabs and move each tab in different directory to run those commands, I want it like when I open windows shell, It automatically open 4 tabs in different file location, Please guide me what Should I do, I tried to use "state" but I guess I applied wrongly or did something wrong. it is not working, 
{
  "$schema": "https://aka.ms/terminal-profiles-schema",
  "actions": [
    {
      "command": "unbound",
      "keys": "ctrl+shift+t"
    },
    {
      "command": "unbound",
      "keys": "ctrl+shift+w"
    },
    {
      "command": "unbound",
      "keys": "alt+f4"
    },
    {
      "command": {
        "action": "copy",
        "singleLine": false
      },
      "keys": "ctrl+c"
    },
    {
      "command": {
        "action": "newTab"
      },
      "keys": "ctrl+t"
    },
    {
      "command": "paste",
      "keys": "ctrl+v"
    },
    {
      "command": "closeWindow"
    },
    {
      "command": "find",
      "keys": "ctrl+shift+f"
    },
    {
      "command": {
        "action": "splitPane",
        "split": "auto",
        "splitMode": "duplicate"
      },
      "keys": "alt+shift+d"
    },
    {
      "command": "closePane",
      "keys": "ctrl+w"
    }
  ],
  "copyFormatting": "none",
  "copyOnSelect": false,
  "defaultProfile": "{61c54bbd-c2c6-5271-96e7-009a87ff44bf}",
  "launchMode": "default",
  "profiles": {
    "defaults": {
      "backgroundImage": null,
      "backgroundImageStretchMode": "uniform",
      "colorScheme": "Solarized Dark",
      "cursorShape": "vintage",
      "font": {
        "face": "Cascadia Code",
        "size": 10,
        "weight": "bold"
      },
      "icon": "C:\\Users\\Azam\\Pictures\\Camera Roll\\extra\\0fMGlH3.jpg",
      "padding": "9",
      "tabTitle": "Welcome Rehan"
    },
    "list": [
      {
        "commandline": "powershell.exe",
        "guid": "{61c54bbd-c2c6-5271-96e7-009a87ff44bf}",
        "hidden": false,
        "name": "Windows PowerShell",
        "colorScheme": "Campbell",
        "cursorColor": "#C19C00",
        "cursorShape": "filledBox",
        "fontSize": 9,
        "padding": "5, 5, 5, 5"
      },
      {
        "commandline": "cmd.exe",
        "guid": "{0caa0dad-35be-5f56-a8ff-afceeeaa6101}",
        "hidden": false,
        "name": "Command Prompt"
      },
      {
        "guid": "{b453ae62-4e3d-5e58-b989-0a998ec441b8}",
        "hidden": false,
        "name": "Azure Cloud Shell",
        "source": "Windows.Terminal.Azure"
      },
      {
        "guid": "{2ece5bfe-50ed-5f3a-ab87-5cd4baafed2b}",
        "hidden": false,
        "name": "Git Bash",
        "source": "Git"
      }
    ]
  },
  "schemes": [
    {
      "background": "#0C0C0C",
      "black": "#000000",
      "blue": "#0037DA",
      "brightBlack": "#767676",
      "brightBlue": "#3B78FF",
      "brightCyan": "#61D6D6",
      "brightGreen": "#16C60C",
      "brightPurple": "#B4009E",
      "brightRed": "#E74856",
      "brightWhite": "#F2F2F2",
      "brightYellow": "#F9F1A5",
      "cursorColor": "#FFFFFF",
      "cyan": "#3A96DD",
      "foreground": "#CCCCCC",
      "green": "#13A10E",
      "name": "Campbell",
      "purple": "#881798",
      "red": "#C50F1F",
      "selectionBackground": "#FFFFFF",
      "white": "#CCCCCC",
      "yellow": "#C19C00"
    }
  ],
  "theme": "system",
  "state": {
    "windows": [
      {
        "maximized": false,
        "positionX": 500,
        "positionY": 200,
        "width": 1000,
        "height": 2000,
        "tabs": [
          {
            "title": "Project X",
            "panes": [
              {
                "profile": "61c54bbd-c2c6-5271-96e7-009a87ff44bf",
                "directory": "C:/Data/Tools/"
              }
            ]
          }
        ]
      }
    ]
  }
}

**Comment by zadjii-msft** [comment]
`state` is not a member in `settings.json`. When I'm talking about `state.json`, I'm talking about a _totally different file_ that the Terminal uses to persist instance data. Setting the global setting [`"firstWindowPreference": "persistedWindowLayout"`](https://docs.microsoft.com/en-us/windows/terminal/customize-settings/startup#behavior-when-starting-a-new-terminal-session-preview) in 1.12 will cause the Terminal to automatically populate your current window state into that file, and re-use it on startup.

Otherwise, what I used to do before that feature landed, was add an action like the following: (copypasta from a while back)

```json
{ "command": { "action": "wt", "commandline": "new-tab --title OpenConsole cmd.exe /k #work 15 ; split-pane -s .30 --title OpenConsole cmd.exe /k #work 15 ; split-pane  -s .25 -H cmd.exe /k media ; new-tab --title \"Symbols Script\" powershell dev\\symbols.ps1 ; new-tab -p \"Ubuntu 18.04\" ; new-tab -p \"microsoft/Terminal\" ; sp -V -p \"microsoft/Terminal\" ; sp -H -p \"microsoft/Terminal\" ; focus-tab -t 0" }, "name": "Good Morning" },
```

(you could probably make it shorter by replacing `new-tab`/`split-pane` with `nt`/`sp`, respectively). (EDIT: You could also probably make this clearer now with the `multipleActions` action)

I add that to the `keybindings`/`actions`. This creates a new command in the [Command Palette](https://docs.microsoft.com/en-us/windows/terminal/command-palette) named "Good Morning". That opens up a few tabs & panes, running various build environments. I like having it in a command rather than `startupActions`, because I only really want this in _one_ terminal window, not _every single one I launch_. It's personal taste. 

You could repeat this for multiple different "session"s if you wanted. That way you could have layouts pre-defined for various different dev environments.

**Comment by theperiscope** [comment]
I'm using Terminal 1.14.2281.0 and would really like the window position to stay persistent but it doesn't. 

I have a `Open windows from a previous session selected'. With that mode, closing my last window with `ALT+F4`, `X`, or `CTRL-SHIFT-W` and then restarting Terminal works as expected and on restart position is restored. However, if you type `exit` (Command Prompt) to close last shell, position does not get restored.

**Comment by Poopooracoocoo** [comment]
I *only* want the window position to be persistent just like every other app. I don't want to restore my previous session at all.

**Comment by zadjii-msft** [comment]
Necro'ing an old thread is not typically the best place for bug reports 😉 However, we definitely do intentionally treat `exit` different than closing the window. If the connection itself initiated the close, then we treat that as a signal that the connection is all done, and doesn't need to come back. Closing from the UI side though does persist the terminal state. We discussed this more in one of #9800, #10972 or #11083, but I can't find the specific comment chain at the moment. 

<hr>
@Poopooracoocoo your issue is already tracked in #12633, no need to pollute discussion here.

**Comment by simontom** [comment]
Well, without preserving the last opened path in each tab it seems pretty useless.
I'm on v1.15.2874
After a few years, this is still no working

**Comment by Kolobamanacas** [comment]
> I apologize if the naming is confusing. The issue, I think, is that a "session" is not well defined yet anywhere, and the wording was based on other applications (such as Edge) to be forward looking for when things like #961 are also implemented.
> 
> The exact behavior of this feature when the setting is enabled is, from the docs (https://docs.microsoft.com/en-us/windows/terminal/customize-settings/startup#behavior-when-starting-a-new-terminal-session-preview)
> 
> ```
> Note: [when enabled] Currently, Windows Terminal will save the following information:
> - Each window's position, size, and name
> - The tab layout of each window, including the layout and profile of each pane, but not any contents of those panes
> ```

@Rosefield 

Sorry for disturbing you in such a manner, but it's been a year and it seems like you have some idea of what's going on with the roadmap. Could you please shed some light on the implementing saving session feature plans?

By saving session I mean the Terminal's ability when launched to reopen tabs with their names, positions and (missed now) the opened paths. Currently I have lets say three projects. I open three PowerShell tabs for these projects and navigate to corresponding path for each of them. But when I reopen Terminal, each tab's path is reset to a user's home directory, which seems like not expecting behavior.

**Comment by zadjii-msft** [comment]
> Well, without preserving the last opened path in each tab it seems pretty useless. I'm on v1.15.2874 After a few years, this is still no working

You know, that's actually in the docs: [Tutorial: Opening a tab or pane in the same directory in Windows Terminal](https://learn.microsoft.com/en-us/windows/terminal/tutorials/new-tab-same-directory). You need to set up your shell for it to cooperate with the Terminal - that doesn't work magically on Windows, unfortunately.

<hr>

@Kolobamanacas This feature was implemented like, a year ago. Did you enable the following setting:
![image](https://user-images.githubusercontent.com/18356694/205655867-5bb75333-8a26-4e3c-94f9-b1ac5569623e.png)


Or do you want a more generic "I want to save this set of tabs as a session, but not necessarily reopen the Terminal with that session every time"? In which case, you might be interested in #11526

**Comment by zadjii-msft** [comment]
![ryan-point](https://user-images.githubusercontent.com/18356694/205948096-c800b823-dc2f-4a3d-92fe-ebe16aa00893.gif)

Okay, this isn't _great_, but I'm gonna put a link to the docs in the settings UI, right by that setting. That'll at least _possibly_ clue someone in that there's more config needed.

**Comment by simontom** [comment]
Well, yes :)
Restore session means restore session. Does not work as the name says. Or we all has just got used to all the stuff been saved when restarting apps ;-)

**Comment by ngocbae98** [comment]
> > > Terminal STILL doesn't persist its window size and position!
> > 
> > 
> > Did you enable the setting? ![image](https://user-images.githubusercontent.com/18356694/138146770-7b4c31e3-586f-4660-bed3-15d4b6d6f5a3.png) It won't work for elevated windows at the moment.
> 
> That's not what I want. That's session restore, right? It looks like what should just be the normal behaviour was tied to that feature.

Thanks

**Comment by ngocbae98** [comment]
> > Well, without preserving the last opened path in each tab it seems pretty useless. I'm on v1.15.2874 After a few years, this is still no working
> 
> You know, that's actually in the docs: [Tutorial: Opening a tab or pane in the same directory in Windows Terminal](https://learn.microsoft.com/en-us/windows/terminal/tutorials/new-tab-same-directory). You need to set up your shell for it to cooperate with the Terminal - that doesn't work magically on Windows, unfortunately.
> 
> @Kolobamanacas This feature was implemented like, a year ago. Did you enable the following setting: ![image](https://user-images.githubusercontent.com/18356694/205655867-5bb75333-8a26-4e3c-94f9-b1ac5569623e.png)
> 
> Or do you want a more generic "I want to save this set of tabs as a session, but not necessarily reopen the Terminal with that session every time"? In which case, you might be interested in #11526

I tried setting it like that, but it only restores the previous number of sessions (number of tabs). But the command content of each tab is completely gone.

**Comment by lhecker** [comment]
@ngocbae98 This is the wrong issue. What you want is #961.
Content restoration is available in v1.21.1272.0 and later which is currently in Windows Terminal Preview. You can find the preview version in the app store or in our release page: https://github.com/microsoft/terminal/releases
